[LV] Add support for partial reductions without a binary op

david-arm opened the pull request:
Consider IR such as this:

for.body:
  %iv = phi i64 [ 0, %entry ], [ %iv.next, %for.body ]
  %accum = phi i32 [ 0, %entry ], [ %add, %for.body ]
  %gep.a = getelementptr i8, ptr %a, i64 %iv
  %load.a = load i8, ptr %gep.a, align 1
  %ext.a = zext i8 %load.a to i32
  %add = add i32 %ext.a, %accum
  %iv.next = add i64 %iv, 1
  %exitcond.not = icmp eq i64 %iv.next, 1025
  br i1 %exitcond.not, label %for.exit, label %for.body

Conceptually we can vectorise this using partial reductions too,
although the current loop vectoriser implementation requires the
accumulation of a multiply. For AArch64 this is easily done with
a udot or sdot with an identity operand, i.e. a vector of (i16 1).

In order to do this I had to teach getScaledReductions that the
accumulated value may come from a unary op, hence there is only
one extension to consider. Similarly, I updated the vplan and
AArch64 TTI cost model to understand the possible unary op.

SamTebbs33 commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -281,31 +281,18 @@ bool VPRecipeBase::isPhi() const {
+  VPRecipeBase *OpR = Op->getDefiningRecipe();
 
   // If the partial reduction is predicated, a select will be operand 0 rather
   // than the binary op
```
binary op -> extend user ?

david-arm replied:
Done

SamTebbs33 reviewed (approved):
LGTM with a couple of nits.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8803,40 +8802,70 @@ bool VPRecipeBuilder::getScaledReductions(
+  std::optional<unsigned> BinOpc;
+  Type *ExtOpTypes[2] = {nullptr};
+
+  auto collectExtInfo = [&Exts,
```
```suggestion
  auto CollectExtInfo = [&Exts,
```

david-arm replied:
Done

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/partial-reduce-dot-product-neon.ll:
```diff
@@ -2108,6 +2108,97 @@ for.exit:                        ; preds = %for.body
   ret i32 %result
 }
 
+define i32 @zext_add_reduc_i8_i32(ptr %a) {
```
should those tests go  in a new file? There's no dot product in the new tests?

david-arm replied:
Done. I've also moved the non-dot-product tests from partial-reduce-dot-product.ll to a new partial-reduce.ll file.

fhahn commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -5046,6 +5056,7 @@ InstructionCost AArch64TTIImpl::getPartialReductionCost(
     if (VFMinValue == Scale)
       return Invalid;
   }
+
```
stray new line.

david-arm replied:
Done

fhahn commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -5024,11 +5024,21 @@ InstructionCost AArch64TTIImpl::getPartialReductionCost(
+                (OpAExtend != OpBExtend && !ST->hasMatMulInt8() &&
+                 !ST->isSVEorStreamingSVEAvailable())))
     return Invalid;
+  assert((BinOp || (OpBExtend == TTI::PR_None && !InputTypeB)) &&
```
The documentation for the interface should probably also be updated, documenting that `Opcode` and the second the second type are optional?

david-arm replied:
Yep, absolutely right! I've tried to amend the documentation for the interface. Please take a look and see if it makes sense.

fhahn commented on llvm/lib/Transforms/Vectorize/VPRecipeBuilder.h:
```diff
@@ -30,18 +30,18 @@ struct VFRange;
 /// accumulator).
 struct PartialReductionChain {
   PartialReductionChain(Instruction *Reduction, Instruction *ExtendA,
-                        Instruction *ExtendB, Instruction *BinOp)
```
Comment above needs updating

david-arm replied:
Done

david-arm commented on llvm/test/Transforms/LoopVectorize/AArch64/partial-reduce-chained.ll:
```diff
@@ -1030,6 +1030,474 @@ for.body:                                         ; preds = %for.body.preheader,
 
+
+; TODO: We probably shouldn't be vectorising this loop with a
+; llvm.experimental.vector.partial.reduce.add.v16i32.v16i32
```
Sorry this TODO was based off the first commit that adds the tests. I'll remove it now.

david-arm committed:
Add tests

david-arm committed:
[LV] Add support for partial reductions without a binary op

Consider IR such as this:

for.body:
  %iv = phi i64 [ 0, %entry ], [ %iv.next, %for.body ]
  %accum = phi i32 [ 0, %entry ], [ %add, %for.body ]
  %gep.a = getelementptr i8, ptr %a, i64 %iv
  %load.a = load i8, ptr %gep.a, align 1
  %ext.a = zext i8 %load.a to i32
  %add = add i32 %ext.a, %accum
  %iv.next = add i64 %iv, 1
  %exitcond.not = icmp eq i64 %iv.next, 1025
  br i1 %exitcond.not, label %for.exit, label %for.body

Conceptually we can vectorise this using partial reductions too,
although the current loop vectoriser implementation requires the
accumulation of a multiply. For AArch64 this is easily done with
a udot or sdot with an identity operand, i.e. a vector of (i16 1).

In order to do this I had to teach getScaledReductions that the
accumulated value may come from a unary op, hence there is only
one extension to consider. Similarly, I updated the vplan and
AArch64 TTI cost model to understand the possible unary op.

Co-authored-by: Matt Devereau <matthew.devereau@arm.com>

david-arm commented:
Gentle ping. :)

fhahn reviewed (commented):
Curious if/how this will interact with https://github.com/llvm/llvm-project/pull/136173? 

Would it potentially help to simplify computeCost & co?

SamTebbs33 commented:
> Curious if/how this will interact with #136173?
> 
> Would it potentially help to simplify computeCost & co?

I think we'll want to turn a partial reduction without a bin op into a `VPExtendedReductionRecipe`  and then lower it back to a partial reduction.

david-arm commented:
> > Curious if/how this will interact with #136173?
> > Would it potentially help to simplify computeCost & co?
> 
> I think we'll want to turn a partial reduction without a bin op into a `VPExtendedReductionRecipe` and then expand it back to a partial reduction.

@fhahn @SamTebbs33 Do you think that https://github.com/llvm/llvm-project/pull/136173 is likely to land soon? Ideally I'd like to progress this patch within the next few weeks. It looks like #136173 also depends upon #113903, which looks like it might still take some time.

david-arm commented:
If both of those PRs are close to landing I could try to check them out and see how my PR interacts with them, but I'd rather do that once I know they're clos(ish) to the finish line.

SamTebbs33 commented:
> > > Curious if/how this will interact with #136173?
> > > Would it potentially help to simplify computeCost & co?
> > 
> > 
> > I think we'll want to turn a partial reduction without a bin op into a `VPExtendedReductionRecipe` and then expand it back to a partial reduction.
> 
> @fhahn @SamTebbs33 Do you think that #136173 is likely to land soon? Ideally I'd like to progress this patch within the next few weeks. It looks like #136173 also depends upon #113903, which looks like it might still take some time.

I think that 136173 is very very close to being approved and there shouldn't be any big changes to it on the way.

MDevereau committed:
run clang-format

MDevereau commented:
Hi, @david-arm asked me to look after this pull request for him.

I've rebased it and have added some new logic to `VPPartialReductionRecipe::computeCost` to handle some situations where different VPlan operands that previously weren't seen are now possible for Partial Reduction recipes.

Regarding the previous conversation about waiting for @SamTebbs33's work, we've discussed the situation internally and would like to prioritize getting this patch along so that it's ready before the branch point for 21.

@fhahn Could you please re-review this patch? Thank you.

MDevereau committed:
s/TargetTransformInfo/TTI, rm TargetScaleFactor change

huntergr-arm reviewed (approved):
LGTM

fhahn reviewed (commented):
Now that the change for bundle/expression recipes landed, https://github.com/llvm/llvm-project/pull/146073 should also be good land soon.

As long as we are confident that https://github.com/llvm/llvm-project/pull/146073 will help simplify the cost implementation we don't necessarily wait for it to land. 

Left a few additional comments inline.

MDevereau commented on llvm/test/Transforms/LoopVectorize/AArch64/partial-reduce.ll:
```diff
@@ -0,0 +1,826 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --check-globals none --filter-out-after "^scalar.ph:" --version 4
```
The test in partial-reduce.ll has the SVE attribute which means it will generate scalable code for the CHECK-MAXBW tests. I can change this one to partial-reduce-sve.ll.

MDevereau committed:
move partial-reduce-neon.ll tests and fix extend comment

MDevereau commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -296,49 +296,70 @@ bool VPRecipeBase::isScalarCast() const {
+  // Pick out opcode, type/ext information and use sub side effects from a widen
+  // recipe.
+  auto HandleWiden = [&](VPWidenRecipe *Widen) {
+    if (match(Widen,
```
When applying this patch on top of #146073, most tests pass even with this entire VPlan being deleted. The main exception I've seen is with the test `zext_add_reduc_i8_i32_has_neon_dotprod` which ends up with an invalid cost as it tries to go through [this](https://github.com/llvm/llvm-project/pull/146073/files#diff-34abe4c3cd34aa7a9664bbd204834248455635ba80b8a9ba9506d8c3e6b94d95R2466-R2471) which isn't passing any information about extends through to getPartialReductionCost

sjoerdmeijer commented:
I am suspecting that this is causing:

https://github.com/llvm/llvm-project/issues/148260
